h2/src/test/org/h2/test/scripts/dml/select.sql

```diff
@@ -803,3 +803,107 @@ SELECT DISTINCT DATA FROM TEST X ORDER BY (CASE WHEN EXISTS(SELECT * FROM TEST T
 
 DROP TABLE TEST;
 > ok
+
+-- Additional GROUP BY tests
+
+CREATE TABLE TEST(A INT, B INT, C INT) AS (VALUES
+    (NULL, NULL, NULL), (NULL, NULL, 1), (NULL, NULL, 2),
+    (NULL, 1, NULL), (NULL, 1, 1), (NULL, 1, 2),
+    (NULL, 2, NULL), (NULL, 2, 1), (NULL, 2, 2),
+    (1, NULL, NULL), (1, NULL, 1), (1, NULL, 2),
+    (1, 1, NULL), (1, 1, 1), (1, 1, 2),
+    (1, 2, NULL), (1, 2, 1), (1, 2, 2),
+    (2, NULL, NULL), (2, NULL, 1), (2, NULL, 2),
+    (2, 1, NULL), (2, 1, 1), (2, 1, 2),
+    (2, 2, NULL), (2, 2, 1), (2, 2, 2));
+> ok
+
+SELECT SUM(A), B, C FROM TEST GROUP BY B, C;
+> SUM(A) B    C
+> ------ ---- ----
+> 3      1    1
+> 3      1    2
+> 3      1    null
+> 3      2    1
+> 3      2    2
+> 3      2    null
+> 3      null 1
+> 3      null 2
+> 3      null null
+> rows: 9
+
+EXPLAIN SELECT SUM(A), B, C FROM TEST GROUP BY B, C;
+>> SELECT SUM("A"), "B", "C" FROM "PUBLIC"."TEST" /* PUBLIC.TEST.tableScan */ GROUP BY "B", "C"
+
+SELECT SUM(A), B, C FROM TEST GROUP BY (B), C, ();
+> SUM(A) B    C
+> ------ ---- ----
+> 3      1    1
+> 3      1    2
+> 3      1    null
+> 3      2    1
+> 3      2    2
+> 3      2    null
+> 3      null 1
+> 3      null 2
+> 3      null null
+> rows: 9
+
+EXPLAIN SELECT SUM(A), B, C FROM TEST GROUP BY (B), C, ();
+>> SELECT SUM("A"), "B", "C" FROM "PUBLIC"."TEST" /* PUBLIC.TEST.tableScan */ GROUP BY "B", "C"
+
+SELECT SUM(A), B, C FROM TEST GROUP BY (B, C);
+> SUM(A) B    C
+> ------ ---- ----
+> 3      1    1
+> 3      1    2
+> 3      1    null
+> 3      2    1
+> 3      2    2
+> 3      2    null
+> 3      null 1
+> 3      null 2
+> 3      null null
+> rows: 9
+
+EXPLAIN SELECT SUM(A), B, C FROM TEST GROUP BY (B, C);
+>> SELECT SUM("A"), "B", "C" FROM "PUBLIC"."TEST" /* PUBLIC.TEST.tableScan */ GROUP BY "B", "C"
+
+SELECT COUNT(*) FROM TEST;
+>> 27
+
+EXPLAIN SELECT COUNT(*) FROM TEST;
+>> SELECT COUNT(*) FROM "PUBLIC"."TEST" /* PUBLIC.TEST.tableScan */ /* direct lookup */
+
+SELECT COUNT(*) FROM TEST GROUP BY ();
```

Comment:
this seems odd? - surely we should not be outputing a GROUP BY clause if there is nothing to group by?

Reply:
There is a problem. In this specific query such clause is useless, it does not affect the execution of the query. But in queries without aggregate functions and without `HAVING` clause (they are perfectly valid) missing `GROUP BY ()` will change execution of the whole query, that's not acceptable.

In my first implementation I made such mistake, but I quickly found this problem.

Of course, we may try to optimize group queries somehow and output empty grouping set only when it really required. But in that case we need to check select expression for presence of window functions and check the presence of `HAVING` clause.

Reply:
BTW, we have another bug.

```SQL
EXPLAIN SELECT FALSE AND MAX(A) > 0 FROM (VALUES 1, 2, 3) T(A);
>> SELECT FALSE FROM (VALUES (1), (2), (3)) "T"("A") /* table scan */ /* direct lookup */
```
Original query returns 1 row, rewritten query (it may happen with subselect or view) returns 3 rows.

To resolve it we should either output `GROUP BY ()` unconditionally (in absence of `HAVING`), or optimized expressions should be somehow scanned for non-window aggregate functions again.

Reply:
Interesting. Also see
    https://blog.jooq.org/2018/05/25/how-to-group-by-nothing-in-sql/
where GROUP BY () changes the output

Reply:
I know that some databases return incorrect results, but it doesn't mean that H2 should do the same.

Reply:
1. Presence of `HAVING` in absence of `GROUP BY ` implies `GROUP BY ()`:
```
1) Let HC be the <having clause>.
Let TE be the <table expression> that immediately contains HC.
If TE does not immediately contain a <group by clause>,
then “GROUP BY ()” is implicit.
```

2. Presence of an aggregate function in absence of `GROUP BY` implies `GROUP BY ()`:
```
16) If <table expression> does not immediately contain a <group by clause>
and <table expression> is simply contained in a <query specification>
that is the aggregation query of some <set function specification>,
then GROUP BY () is implicit.
```

So `GROUP BY ()` and missing `GROUP BY` clauses in a grouped query are equivalent.

Reply:
BTW, Oracle 11g R2 really has such bug, but Oracle 18c returns correct grand total as expected, I don't know the exact version when it was fixed, but it doesn't matter. DB2 Developer-C 11.1 also returns correct result. SQL Server is still affected.